[requires] add tests for requires.io service, review error handling

chris48s opened the pull request.

paulmelnikow commented on service-tests/requires.js:
```diff
@@ -0,0 +1,41 @@
+    .get('/api/v1/status/github/celery/celery')
+    .reply(200, "{{{{{invalid json}}")
+  )
+  .expectJSON({name: 'requirements', value: 'unknown'});
```
This is the same message as the 404 case. Could you change this one to `'invalid'` and the 404 case to `'not found'`?

chris48s commented:
> This is the same message as the 404 case.

Given this situation is going to come up a lot (perhaps even hundreds of times), its probably worth zooming out a little bit and thinking about this in the slightly wider context of The Great Rewrite™ :)

If I'm writing a test and I find a badge which is doing something obviously wrong, there is absolutely no point writing a test that formalises a bug. i.e: if I'm finding a badge that does this

* [![badge](https://img.shields.io/github/issues/foobar/not-a-repo.svg)]() - https://img.shields.io/github/issues/foobar/not-a-repo.svg

or this

* [![badge](https://img.shields.io/github/contributors/foobar/not-a-repo.svg)]() - https://img.shields.io/github/contributors/foobar/not-a-repo.svg

then part of the process of testing that should indisputably be to fix the bug.

Another very common scenario (probably the most common) is that badges handle the 'not found' case by returning a generic error response that is not specific to 'not found' but which may also be returned in other error cases. This is very common, even in code which already has service tests. For example:

* [![badge](https://img.shields.io/travis/rust-lang/not-a-repo.svg)]() - https://img.shields.io/travis/rust-lang/not-a-repo.svg
* [![badge](https://img.shields.io/luarocks/v/mpeterv/not-a-package.svg)]() - https://img.shields.io/luarocks/v/mpeterv/not-a-package.svg
* [![badge](https://img.shields.io/maven-central/v/org.apache.maven/not-a-package.svg)]() - https://img.shields.io/maven-central/v/org.apache.maven/not-a-package.svg

(all of the examples above are from integrations with an existing suite of tests)

As you've noted in https://github.com/badges/shields/issues/1358#issuecomment-350377904 very few badges have implemented a specific 'not found' path, so this case will occur in most of the integrations we've identified as untested and also in many of the ones that already have tests.

Obviously 'not found' is a particular error case which is relevant to most of the service integrations and The Great Rewrite™ is an opportunity to try and handle this as a piece of 'boilerplate' that can be abstracted. It is also an opportunity to 'standardise' because it is something that is handled in different ways, even where it is handled.

Given that context, I can see 2 ways of looking at this:

1. We say that specific handling for 'not found' is something that should be done for every integration in the current `server.js` structure as part of the process of adding tests. Not implementing a specific error case for 'not found' should be considered a bug (in the same way that returning `NaN` or `undefined` in a badge is) and if we find it while adding tests, it should be fixed as part of the process of testing.

2. Alternatively, we say that if 'not found' is handled by returning some kind of error response (even if that error is also returned in other invalid/error cases), its fine to fomalise that with a test. Specific error cases for 'not found' can be handled as part of the process of porting to a new structure.

I can see reasons for both approaches and I'm not strongly advocating for either (although as you can see, I've been leaning towards the second), but I think it would be useful to pick one approach or the other now, document it, and do it consistently moving forwards. This will ensure we don't end up with too much of a patchwork and ensure everyone has consistent expectations when submitting or reviewing PRs.

Thoughts?

paulmelnikow commented:
Hey @chris48s, this is a really thoughtful question! It's been a busy week. I've started putting some thoughts together but haven't had time to assemble them into a response. Will do that soon!

platan commented:
When it comes to 'not found' case maybe we are able to write one test for all bagdes (or at least for a large part of them). Such test could create badge for non-existing resource (package/repo). Handling response status code 404 in abstracted part of the code should be sufficient for most of the badges. With this approach we would have one test for 'not found' case. 

@chris48s 
Rewriting all services could take some time and improving them right now seems to be valuable. But bearing that we want to rewrite services entirely it's not worth to invest time in handling 'not found'.

paulmelnikow commented:
> Obviously 'not found' is a particular error case which is relevant to most of the service integrations and The Great Rewrite™ is an opportunity to try and handle this as a piece of 'boilerplate' that can be abstracted. It is also an opportunity to 'standardise' because it is something that is handled in different ways, even where it is handled.

Definitely agreed.

> When it comes to 'not found' case maybe we are able to write one test for all badges (or at least for a large part of them). Such test could create badge for non-existing resource (package/repo). Handling response status code 404 in abstracted part of the code should be sufficient for most of the badges. With this approach we would have one test for 'not found' case.

Yea! I’m excited about the prospect of being able to refactor across the new services. If we have 100% coverage of all the new services, this becomes easy and fun.

It does seem like we’ll be able to generalize a lot of the error handling.

From our perspective, 404 is the correct thing for a service to return for a nonexistent entity. Unfortunately they don’t all do this, which is why detecting this case requires case-by-case work. Some services return empty JSON, differently structured JSON, or a badge with error text on it. I don't think even a majority of services return a 404.

I imagine _our_ implementation for "not found" will involve throwing a NotFound exception, or perhaps returning a NotFound object, which the calling code can check explicitly. The badge message could be set by the badge, and default to “not found,” or perhaps something a bit more pedantic like “no such entity.”

> Given that context, I can see 2 ways of looking at this:

I’m in favor of approach (1) you outlined, and here’s why. Writing a good validation regex for the success cases involve understanding the service, and often, in my experience, reading the API docs. This activity is closely coupled with the service itself. Writing tests sometimes means finding real projects to cover all the important cases. Determining how “not found” is represented by the API requires similar work – looking at responses, API docs, and such.

In the presence of complete tests, The Great Rewrite (I love it – it’s catching on!) may not require interacting with the service at all. Ideally the work is purely implementation driven. New code that matches the old behavior and can be completely validated with existing tests.

So having written a few services from scratch, and reviewed a lot of them, my instinct is that doing the 404-detection work is a good investment, creating value which carries over post-rewrite.

What do you think?

chris48s commented:
OK cool. Thanks for responding. Its been a couple of weeks since I did anything on this now so I've got a bit out of the groove. Hopefully I will have a bit of time clear over the holidays to get back into it and start progressing based on this feedback. Obviously I'll whip this and #1362 into shape but it is probably also worth looking at whether I can do anything to make this a bit more standardized/repeatable given it is something that will need to be done in quite a few places.
Cheers

paulmelnikow commented:
Yea, I understand about that. Sorry for my slow response – will try to be more responsive in the future!

chris48s commented:
No worries. We all have lives outside of open source :) :christmas_tree:

chris48s commented:
I've updated the PR to demonstrate where I'm thinking of going with this as an intermediate position.

Obviously this will not cover all cases. In a situation where the 'not found' response isn't a 404, we will need to define custom error handling for that particular badge. In cases which do follow this pattern, we can use this to abstract a bit of fairly common boilerplate.

I'd like to take this further and refactor more badges which already define explicit handlers and have tests (as I've done in ea008ad, ce502a7, 06a6c04 ) but before I go further and do that it would be useful to confirm you're happy with this approach. Assuming that's the case I will also add some notes to the docs.

Note: sorry - I screwed up naming the PR first time so the build failed. I've renamed it so if you restart the build, the service tests should pass..

paulmelnikow commented:
Build came out green after restart. 😄 

I'm 100% for generalizing the error handling, and really like the approach of implementing reusable pieces like you've done here. However I've been pushing back on refactoring `server.js` in favor of handling changes like this as part of the rewrite. For the purpose of the rewrite, I'm basically thinking of `server.js` as throwaway. There may be snippets of useful expressions we can copy and paste (e.g. how to build service URLs) and useful logic (e.g. how to construct how to identify the not-found case), though I expect most of the code to be used as a reference and then discarded. The main reason is the change in concurrency pattern produces really dramatic changes in the code, both in the service and in the helpers themselves. It's especially true of error handling: being able to bubble a NotFound exception up to the BaseService will get rid of so many boilerplate conditionals.

My preference would be to make minimal changes to the service since they're easiest to read and will be thrown away soon. Since I'm asking for functionality changes as the tests are written, could I suggest a compromise? How about we continue to refactor the error handling as you've done here, but only where (1) we need to add code to detect not found and (2) the check in this function works (i.e. the service returns 404).

What do you think?

chris48s commented:
OK no worries. I've backed out those additional commits. Review this as it stands and we can just look at refactoring other service code as we need to modify it.

paulmelnikow commented on lib/error-helper.js:
```diff
@@ -0,0 +1,26 @@
+'use strict';
+
+const getStandardErrorResponse = function(err, res) {
```
What do you think about tweaking this?

```js
function checkResponseStatus(badgeData, err, res) {
    if (err != null) {
        badgeData.text[1] = 'inaccessible';
        badgeData.colorscheme = 'red';
        return true;
    }
    ...
    return false;
}
```

Then it could be invoked more tersely:

```js
if (checkResponseStatus(badgeData, err, res)) {
    sendBadge(format, badgeData);
    return;
}
```

This version is slightly harder to test, but seems worth it.

chris48s commented:
I've added a commit that implements your suggestion but I've decided to call it the function `checkErrorResponse()` because a function called `checkResponseStatus()` which returns `true` when there is an error and `false` when everything is ok seems like it is the wrong way round.

I think in general it is nicer to use a pure function instead of side effects, but as you note it is essentially a temporary solution and we will end up using this snippet in a lot of cases so making it more terse to invoke here is probably a valid tradeoff.

As it stands, if the function were invoked with a `badgeData` param that does not have a `.text` property (e.g: `checkErrorResponse({},...)` ) an exception will be thrown trying to set `.text[1]`. Do you think we should try to cover this and return a valid badge object in this case, or should it be the responsibility of the calling code to invoke with a valid badge object?

paulmelnikow commented:
Sounds good.

> As it stands, if the function were invoked with a `badgeData` param that does not have a `.text` property (e.g: `checkErrorResponse({},...)` ) an exception will be thrown trying to set `.text[1]`. Do you think we should try to cover this and return a valid badge object in this case, or should it be the responsibility of the calling code to invoke with a valid badge object?

It should be the responsibility of the calling code to invoke with a valid badge object.

All the `badgeData` are created by `getBadgeData()` which has a `text` array, and as you point out it's a temporary solution.

paulmelnikow commented:
Thanks a bunch! 🏅
